pageserver/src/tenant/timeline/init.rs

```diff
@@ -0,0 +1,207 @@
+                    } else {
+                        Discovered::Unknown(file_name)
+                    }
+                }
+            }
+        };
+
+        ret.push(discovered);
+    }
+
+    Ok(ret)
+}
+
+/// Decision on what to do with a layer file after considering it's local and remote metadata.
+#[derive(Clone)]
+pub(super) enum Decision {
+    /// The layer is not present locally.
+    Evicted(LayerFileMetadata),
+    /// The layer is present locally, but local metadata does not match remote; we must
+    /// delete it and treat it as evicted.
+    UseRemote {
+        local: LayerFileMetadata,
+        remote: LayerFileMetadata,
+    },
+    /// The layer is present locally, and metadata matches.
+    UseLocal(LayerFileMetadata),
+    /// The layer is only known locally, it needs to be uploaded.
+    NeedsUpload(LayerFileMetadata),
+}
+
+/// The related layer is is in future compared to disk_consistent_lsn, it must not be loaded.
+#[derive(Debug)]
+pub(super) struct FutureLayer;
+
+/// Merges local discoveries and remote [`IndexPart`] to a collection of decisions.
+pub(super) fn recoincile(
+    discovered: Vec<(LayerFileName, u64)>,
+    index_part: Option<&IndexPart>,
+    disk_consistent_lsn: Lsn,
+) -> Vec<(LayerFileName, Result<Decision, FutureLayer>)> {
+    use Decision::*;
+    use LayerFileName::*;
+
+    // name => (local, remote)
+    type Collected = HashMap<LayerFileName, (Option<LayerFileMetadata>, Option<LayerFileMetadata>)>;
+
+    let mut discovered = discovered
+        .into_iter()
+        .map(|(name, file_size)| (name, (Some(LayerFileMetadata::new(file_size)), None)))
+        .collect::<Collected>();
+
+    // merge any index_part information, if we would have such
+    index_part
+        .as_ref()
+        .map(|ip| ip.layer_metadata.iter())
+        .into_iter()
+        .flatten()
+        .map(|(name, metadata)| (name, LayerFileMetadata::from(metadata)))
+        .for_each(|(name, metadata)| {
+            if let Some(existing) = discovered.get_mut(name) {
+                existing.1 = Some(metadata);
+            } else {
+                discovered.insert(name.to_owned(), (None, Some(metadata)));
+            }
+        });
+
+    // must not use the index_part presence here as a proxy for remote timeline client
+    // it might be that we have local file(s), but the index upload was stopped with failpoint
+
+    discovered
+        .into_iter()
+        .map(|(name, (local, remote))| {
+            let future = match &name {
+                Image(file_name) if file_name.lsn > disk_consistent_lsn => true,
+                Delta(file_name) if file_name.lsn_range.end > disk_consistent_lsn + 1 => true,
+                _ => false,
+            };
+
+            let decision = if future {
+                Err(FutureLayer)
+            } else {
+                Ok(match (local, remote) {
+                    (Some(local), Some(remote)) if local != remote => UseRemote { local, remote },
```

Comment:
So, if we produce an equally named layer with different contents locally, remote wins?
That sounds
1. useful to recover from local bitflips (currently we just check file size but we both know it's going to be a checksum in the future)
2. not good for the case where we locally produced the same layer file name but with different contents. Can that happen or is the `merge_local_and_remote_metadata` somewhere in the callers taking care of that condition earlier, making sure local wins in those cases?

Reply:
I don't think this is changed behaviour: https://github.com/neondatabase/neon/blob/71ccb07a43fec68e203ff1dc09608c8d108f21f3/pageserver/src/tenant/timeline.rs#L1775C27-L1785

Note the metadata here is LayerFileMetadata which is an u64 or a "value type which has the aspects we care about".

In remote_timeline_client module comment there's an assumption of immutability of the files. The fact that we sometimes overwrite is a bug, and no one knows about the contents. Currently we hope we'll get to upload the file.

Reply:
> In remote_timeline_client module comment there's an assumption of immutability of the files. The fact that we sometimes overwrite is a bug, and no one knows about the contents. Currently we hope we'll get to upload the file.

It's more subtle.
We currently never overwrite / re-use layer file names because
1. we rename-to-backup layers from the future, so, we'd rather produce an L0 layer again by re-ingesting WAL, than overwriting it in some way
2. for compaction, where we just rearrange the past, I guess there are some bugs due to the fact that compaction isn't atomic wrt timeline dir.

The definition of "what is future" is given by `merge_local_and_remote_metadata`, at least that's how it is in the old code.
`merge_local_and_remote_metadata` hard-codes the assumption that local disk consistent lsn is >= remote consistent lsn, and bails out if that's not the case.
AFAICT that code didn't change.

Reply:
I think we do exactly overwrite L1's, we do it silently as far as I remember, see #4088 and what I just created #5077.

> AFAICT that code didn't change.

No.

All in all, I don't see a problem here.

Reply:
Yeah, my doubts are cleared, thanks for creating #5077